File: android/ssabree/app/src/main/java/com/ssafy/ssabree/features/findidpass/viewmodel/FindIdPassViewModel.kt
package com.ssafy.ssabree.features.findidpass.viewmodel

import android.util.Log
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.ssafy.ssabree.core.repository.AuthRepository
import kotlinx.coroutines.launch

private const val TAG = "FindIdPassViewModel"
private const val CODE_LENGTH = 6

data class FindIdPassUiState(
    val findIdMattermostId: String = "",
    val findIdGeneration: Int? = null,
    val findIdName: String = "",
    val findIdVerificationCode: String = "",
    val isFindIdGenerationDropdownExpanded: Boolean = false,
    val isFindIdCodeSent: Boolean = false,
    val isFindIdLoading: Boolean = false,
    val findIdError: String? = null,
    val foundEmail: String? = null,

    val findPassMattermostId: String = "",
    val findPassGeneration: Int? = null,
    val findPassName: String = "",
    val findPassVerificationCode: String = "",
    val isFindPassGenerationDropdownExpanded: Boolean = false,
    val isFindPassCodeSent: Boolean = false,
    val isFindPassVerified: Boolean = false,
    val isFindPassLoading: Boolean = false,
    val findPassError: String? = null,

    val newPassword: String = "",
    val newPasswordCheck: String = "",
    val isResetLoading: Boolean = false,
    val isResetSuccess: Boolean = false,
)

class FindIdPassViewModel(
    private val authRepository: AuthRepository
) : ViewModel() {

    private fun normalizeMattermostId(value: String): String {
        return value.trim().removePrefix("@")
    }

    var uiState by mutableStateOf(FindIdPassUiState())
        private set

    fun onFindIdMattermostIdChange(value: String) {
        uiState = uiState.copy(findIdMattermostId = value, findIdError = null)
    }

    fun onFindIdNameChange(value: String) {
        uiState = uiState.copy(findIdName = value, findIdError = null)
    }

    fun onFindIdGenerationDropdownClick() {
        uiState = uiState.copy(isFindIdGenerationDropdownExpanded = true)
    }

    fun onFindIdGenerationDropdownDismiss() {
        uiState = uiState.copy(isFindIdGenerationDropdownExpanded = false)
    }

    fun onFindIdGenerationSelected(value: Int) {
        uiState = uiState.copy(
            findIdGeneration = value,
            isFindIdGenerationDropdownExpanded = false,
            findIdError = null
        )
    }

    fun onFindIdVerificationCodeChange(value: String) {
        uiState = uiState.copy(
            findIdVerificationCode = value.filter { it.isDigit() }.take(CODE_LENGTH),
            findIdError = null
        )
    }

    fun sendFindIdCode() {
        val mattermostId = normalizeMattermostId(uiState.findIdMattermostId)
        val generation = uiState.findIdGeneration
        val name = uiState.findIdName
        if (name.isBlank() || mattermostId.isBlank() || generation == null) {
            uiState = uiState.copy(findIdError = "이름, Mattermost ID와 기수를 입력해주세요.")
            return
        }

        viewModelScope.launch {
            uiState = uiState.copy(isFindIdLoading = true, findIdError = null)
            authRepository.requestSsafyVerification(mattermostId, generation, name)
                .onSuccess {
                    Log.d(TAG, "sendFindIdCode: succeed")
                    uiState = uiState.copy(isFindIdLoading = false, isFindIdCodeSent = true)
                }
                .onFailure { e ->
                    Log.d(TAG, "sendFindIdCode: failed (${e.message})")
                    uiState = uiState.copy(isFindIdLoading = false, findIdError = e.message)
                }
        }
    }

    fun verifyFindId() {
        val mattermostId = normalizeMattermostId(uiState.findIdMattermostId)
        val code = uiState.findIdVerificationCode
        if (mattermostId.isBlank() || code.length != CODE_LENGTH) return

        viewModelScope.launch {
            uiState = uiState.copy(isFindIdLoading = true, findIdError = null)
            authRepository.confirmSsafyVerification(mattermostId, code)
                .onSuccess {
                    Log.d(TAG, "verifyFindId - confirmSsafyVerification: succeed")
                    authRepository.findId(mattermostId)
                        .onSuccess { email ->
                            Log.d(TAG, "verifyFindId - findId: succeed")
                            uiState = uiState.copy(
                                isFindIdLoading = false,
                                foundEmail = email
                            )
                        }
                        .onFailure { e ->
                            Log.d(TAG, "verifyFindId - findId: failed (${e.message})")
                            uiState = uiState.copy(
                                isFindIdLoading = false,
                                findIdError = e.message
                            )
                        }
                }
                .onFailure { e ->
                    Log.d(TAG, "verifyFindId - confirmSsafyVerification: failed (${e.message})")
                    uiState = uiState.copy(
                        isFindIdLoading = false,
                        findIdError = e.message
                    )
                }
        }
    }

    fun onFindPassMattermostIdChange(value: String) {
        uiState = uiState.copy(findPassMattermostId = value, findPassError = null)
    }

    fun onFindPassNameChange(value: String) {
        uiState = uiState.copy(findPassName = value, findPassError = null)
    }

    fun onFindPassGenerationDropdownClick() {
        uiState = uiState.copy(isFindPassGenerationDropdownExpanded = true)
    }

    fun onFindPassGenerationDropdownDismiss() {
        uiState = uiState.copy(isFindPassGenerationDropdownExpanded = false)
    }

    fun onFindPassGenerationSelected(value: Int) {
        uiState = uiState.copy(
            findPassGeneration = value,
            isFindPassGenerationDropdownExpanded = false,
            findPassError = null
        )
    }

    fun onFindPassVerificationCodeChange(value: String) {
        uiState = uiState.copy(
            findPassVerificationCode = value.filter { it.isDigit() }.take(CODE_LENGTH),
            findPassError = null
        )
    }

    fun sendFindPassCode() {
        val mattermostId = normalizeMattermostId(uiState.findPassMattermostId)
        val generation = uiState.findPassGeneration
        val name = uiState.findPassName
        if (name.isBlank() || mattermostId.isBlank() || generation == null) {
            uiState = uiState.copy(findPassError = "이름, Mattermost ID와 기수를 입력해주세요.")
            return
        }

        viewModelScope.launch {
            uiState = uiState.copy(isFindPassLoading = true, findPassError = null)
            authRepository.requestSsafyVerification(mattermostId, generation, name)
                .onSuccess {
                    Log.d(TAG, "sendFindPassCode: succeed")
                    uiState = uiState.copy(isFindPassLoading = false, isFindPassCodeSent = true)
                }
                .onFailure { e ->
                    Log.d(TAG, "sendFindPassCode: failed (${e.message})")
                    uiState = uiState.copy(isFindPassLoading = false, findPassError = e.message)
                }
        }
    }

    fun verifyFindPass() {
        val mattermostId = normalizeMattermostId(uiState.findPassMattermostId)
        val code = uiState.findPassVerificationCode
        if (mattermostId.isBlank() || code.length != CODE_LENGTH) return

        viewModelScope.launch {
            uiState = uiState.copy(isFindPassLoading = true, findPassError = null)
            authRepository.confirmSsafyVerification(mattermostId, code)
                .onSuccess {
                    Log.d(TAG, "verifyFindPass: succeed")
                    uiState = uiState.copy(
                        isFindPassLoading = false,
                        isFindPassVerified = true
                    )
                }
                .onFailure { e ->
                    Log.d(TAG, "verifyFindPass: failed (${e.message})")
                    uiState = uiState.copy(
                        isFindPassLoading = false,
                        findPassError = e.message
                    )
                }
        }
    }

    fun onNewPasswordChange(value: String) {
        uiState = uiState.copy(newPassword = value, findPassError = null)
    }

    fun onNewPasswordCheckChange(value: String) {
        uiState = uiState.copy(newPasswordCheck = value, findPassError = null)
    }

    fun resetPassword() {
        val mattermostId = normalizeMattermostId(uiState.findPassMattermostId)
        val newPassword = uiState.newPassword
        val newPasswordCheck = uiState.newPasswordCheck

        if (mattermostId.isBlank() || newPassword.isBlank() || newPasswordCheck.isBlank()) return
        if (newPassword != newPasswordCheck) {
            uiState = uiState.copy(findPassError = "비밀번호가 일치하지 않습니다.")
            return
        }

        viewModelScope.launch {
            uiState = uiState.copy(isResetLoading = true, findPassError = null)
            authRepository.resetPassword(mattermostId, newPassword)
                .onSuccess {
                    Log.d(TAG, "resetPassword: succeed")
                    uiState = uiState.copy(
                        isResetLoading = false,
                        isResetSuccess = true
                    )
                }
                .onFailure { e ->
                    Log.d(TAG, "resetPassword: failed (${e.message})")
                    uiState = uiState.copy(
                        isResetLoading = false,
                        findPassError = e.message
                    )
                }
        }
    }

    fun consumeFoundEmail() {
        uiState = uiState.copy(foundEmail = null)
    }

    fun consumeResetSuccess() {
        uiState = uiState.copy(isResetSuccess = false)
    }

    fun consumeFindPassVerified() {
        uiState = uiState.copy(isFindPassVerified = false)
    }
}

File: iOS/ssabree/ssabree/Features/FindIdPass/FindIdPassViewModel.swift
import SwiftUI
import Observation

// MARK: - FindIdPass UI State
// Android FindIdPassUiState와 동일한 구조

struct FindIdPassUiState {
    // 아이디 찾기
    var findIdMattermostId: String = ""
    var findIdGeneration: Int? = nil
    var findIdName: String = ""
    var findIdVerificationCode: String = ""
    var isFindIdGenerationDropdownExpanded: Bool = false
    var isFindIdCodeSent: Bool = false
    var isFindIdLoading: Bool = false
    var findIdError: String? = nil
    var foundEmail: String? = nil

    // 비밀번호 찾기
    var findPassMattermostId: String = ""
    var findPassGeneration: Int? = nil
    var findPassName: String = ""
    var findPassVerificationCode: String = ""
    var isFindPassGenerationDropdownExpanded: Bool = false
    var isFindPassCodeSent: Bool = false
    var isFindPassVerified: Bool = false
    var isFindPassLoading: Bool = false
    var findPassError: String? = nil

    // 비밀번호 재설정
    var newPassword: String = ""
    var newPasswordCheck: String = ""
    var isResetLoading: Bool = false
    var isResetSuccess: Bool = false
}

// MARK: - FindIdPass ViewModel

private let CODE_LENGTH = 6

@MainActor
@Observable
final class FindIdPassViewModel {
    var uiState = FindIdPassUiState()

    private let authRepository: AuthRepository

    init(authRepository: AuthRepository) {
        self.authRepository = authRepository
    }

    // MARK: - Helpers

    private func normalizeMattermostId(_ value: String) -> String {
        var result = value.trimmingCharacters(in: .whitespacesAndNewlines)
        if result.hasPrefix("@") {
            result.removeFirst()
        }
        return result
    }

    // MARK: - 아이디 찾기

    func onFindIdMattermostIdChange(_ value: String) {
        uiState.findIdMattermostId = value
        uiState.findIdError = nil
    }

    func onFindIdNameChange(_ value: String) {
        uiState.findIdName = value
        uiState.findIdError = nil
    }

    func onFindIdGenerationDropdownClick() {
        uiState.isFindIdGenerationDropdownExpanded = true
    }

    func onFindIdGenerationDropdownDismiss() {
        uiState.isFindIdGenerationDropdownExpanded = false
    }

    func onFindIdGenerationSelected(_ value: Int) {
        uiState.findIdGeneration = value
        uiState.isFindIdGenerationDropdownExpanded = false
        uiState.findIdError = nil
    }

    func onFindIdVerificationCodeChange(_ value: String) {
        uiState.findIdVerificationCode = String(value.filter { $0.isNumber }.prefix(CODE_LENGTH))
        uiState.findIdError = nil
    }

    func sendFindIdCode() {
        let mattermostId = normalizeMattermostId(uiState.findIdMattermostId)
        let name = uiState.findIdName
        guard let generation = uiState.findIdGeneration else {
            uiState.findIdError = "이름, Mattermost ID와 기수를 입력해주세요."
            return
        }
        guard !name.isEmpty, !mattermostId.isEmpty else {
            uiState.findIdError = "이름, Mattermost ID와 기수를 입력해주세요."
            return
        }

        Task {
            uiState.isFindIdLoading = true
            uiState.findIdError = nil

            let result = await authRepository.requestSsafyVerification(
                mattermostId: mattermostId,
                generation: generation,
                name: name
            )

            switch result {
            case .success:
                uiState.isFindIdLoading = false
                uiState.isFindIdCodeSent = true
            case .failure(let error):
                uiState.isFindIdLoading = false
                uiState.findIdError = error.localizedDescription
            }
        }
    }

    func verifyFindId() {
        let mattermostId = normalizeMattermostId(uiState.findIdMattermostId)
        let code = uiState.findIdVerificationCode
        guard !mattermostId.isEmpty, code.count == CODE_LENGTH else { return }

        Task {
            uiState.isFindIdLoading = true
            uiState.findIdError = nil

            let verifyResult = await authRepository.confirmSsafyVerification(
                mattermostId: mattermostId,
                code: code
            )

            switch verifyResult {
            case .success:
                // 인증 성공 후 아이디 조회
                let findIdResult = await authRepository.findId(mattermostId: mattermostId)
                switch findIdResult {
                case .success(let email):
                    uiState.isFindIdLoading = false
                    uiState.foundEmail = email
                case .failure(let error):
                    uiState.isFindIdLoading = false
                    uiState.findIdError = error.localizedDescription
                }
            case .failure(let error):
                uiState.isFindIdLoading = false
                uiState.findIdError = error.localizedDescription
            }
        }
    }

    func consumeFoundEmail() {
        uiState.foundEmail = nil
    }

    // MARK: - 비밀번호 찾기

    func onFindPassMattermostIdChange(_ value: String) {
        uiState.findPassMattermostId = value
        uiState.findPassError = nil
    }

    func onFindPassNameChange(_ value: String) {
        uiState.findPassName = value
        uiState.findPassError = nil
    }

    func onFindPassGenerationDropdownClick() {
        uiState.isFindPassGenerationDropdownExpanded = true
    }

    func onFindPassGenerationDropdownDismiss() {
        uiState.isFindPassGenerationDropdownExpanded = false
    }

    func onFindPassGenerationSelected(_ value: Int) {
        uiState.findPassGeneration = value
        uiState.isFindPassGenerationDropdownExpanded = false
        uiState.findPassError = nil
    }

    func onFindPassVerificationCodeChange(_ value: String) {
        uiState.findPassVerificationCode = String(value.filter { $0.isNumber }.prefix(CODE_LENGTH))
        uiState.findPassError = nil
    }

    func sendFindPassCode() {
        let mattermostId = normalizeMattermostId(uiState.findPassMattermostId)
        let name = uiState.findPassName
        guard let generation = uiState.findPassGeneration else {
            uiState.findPassError = "이름, Mattermost ID와 기수를 입력해주세요."
            return
        }
        guard !name.isEmpty, !mattermostId.isEmpty else {
            uiState.findPassError = "이름, Mattermost ID와 기수를 입력해주세요."
            return
        }

        Task {
            uiState.isFindPassLoading = true
            uiState.findPassError = nil

            let result = await authRepository.requestSsafyVerification(
                mattermostId: mattermostId,
                generation: generation,
                name: name
            )

            switch result {
            case .success:
                uiState.isFindPassLoading = false
                uiState.isFindPassCodeSent = true
            case .failure(let error):
                uiState.isFindPassLoading = false
                uiState.findPassError = error.localizedDescription
            }
        }
    }

    func verifyFindPass() {
        let mattermostId = normalizeMattermostId(uiState.findPassMattermostId)
        let code = uiState.findPassVerificationCode
        guard !mattermostId.isEmpty, code.count == CODE_LENGTH else { return }

        Task {
            uiState.isFindPassLoading = true
            uiState.findPassError = nil

            let result = await authRepository.confirmSsafyVerification(
                mattermostId: mattermostId,
                code: code
            )

            switch result {
            case .success:
                uiState.isFindPassLoading = false
                uiState.isFindPassVerified = true
            case .failure(let error):
                uiState.isFindPassLoading = false
                uiState.findPassError = error.localizedDescription
            }
        }
    }

    func consumeFindPassVerified() {
        uiState.isFindPassVerified = false
    }

    // MARK: - 비밀번호 재설정

    func onNewPasswordChange(_ value: String) {
        uiState.newPassword = value
        uiState.findPassError = nil
    }

    func onNewPasswordCheckChange(_ value: String) {
        uiState.newPasswordCheck = value
        uiState.findPassError = nil
    }

    func resetPassword() {
        let mattermostId = normalizeMattermostId(uiState.findPassMattermostId)
        let newPassword = uiState.newPassword
        let newPasswordCheck = uiState.newPasswordCheck

        guard !mattermostId.isEmpty, !newPassword.isEmpty, !newPasswordCheck.isEmpty else { return }
        guard newPassword == newPasswordCheck else {
            uiState.findPassError = "비밀번호가 일치하지 않습니다."
            return
        }

        Task {
            uiState.isResetLoading = true
            uiState.findPassError = nil

            let result = await authRepository.resetPassword(
                mattermostId: mattermostId,
                newPassword: newPassword
            )

            switch result {
            case .success:
                uiState.isResetLoading = false
                uiState.isResetSuccess = true
            case .failure(let error):
                uiState.isResetLoading = false
                uiState.findPassError = error.localizedDescription
            }
        }
    }

    func consumeResetSuccess() {
        uiState.isResetSuccess = false
    }
}
